fix 'depend' target in otherlibs/bigarray

gasche opened the pull request:
This target has been broken by the move of bigarray C files to the runtime.
Before the current PR:

```
$ make depend
gcc -MM -O2 -fno-strict-aliasing -fwrapv -Wall -Werror -fno-tree-vrp -fPIC -I../unix -DIN_OCAML_BIGARRAY -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -I../../byterun *.c | sed -e 's/\.o/.$(O)/g' > .depend
gcc: error: *.c: No such file or directory
gcc: fatal error: no input files
```

xclerc commented:
If there are no C files in the directory, does the condition
over MSVC still make sense?

gasche committed:
fix 'depend' target in otherlibs/bigarray

This target has been broken by the move of bigarray C files to the runtime.
Before the current PR:

```
$ make depend
gcc -MM -O2 -fno-strict-aliasing -fwrapv -Wall -Werror -fno-tree-vrp -fPIC -I../unix -DIN_OCAML_BIGARRAY -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -I../../byterun *.c | sed -e 's/\.o/.$(O)/g' > .depend
gcc: error: *.c: No such file or directory
gcc: fatal error: no input files
```

gasche commented:
Good point @xclerc, I just removed it. Thanks!

shindere commented:
It's weird, Inria's CI is supposed to check that alldepend works. I am
owndering how it's possible that it didn't detect this problem as it
appeared.

xclerc commented:
@shindere My understanding is that the pipe to `sed` causes
the whole thing to return an exit code of `0`. The message
reproduced by @gasche is present in the output but the build
does not fail.

shindere commented:
OK I now understand why `make alldepend`does not fail. It's because the
output of gcc is piped through sed and the return code for the pipeline is 0
even if gcc fails.

I guess we could use the pipefail option to fix this but IIRC that woul not
be portable to e.g. dash and Posix shells.

One possibility to fix this in a portalbe way would be to write the output
of gcc directly to the `.depend` file and then use `sed -i` on it, or one
could write to an auxiliary file but one would then have to remove it.

Shall I do a PR that uses `sed -i`?

xclerc commented:
On the topic of CI, wouldn't it be useful to also
ensure that the execution of `make alldepend`
does not produce any change?

shindere commented:
IMO, feel free to merge and I guess this should be cherryp-icked to 4.07, too

shindere commented:
Xavier Clerc (2018/06/19 13:47 +0000):
> On the topic of CI, wouldn't it be useful to also
> ensure that the execution of `make alldepend`
> does not produce any change?

Yeah I considered that, a long time ago.

I renounced because there may be false positives depending on the
version of gcc which has been used to update the dependencies.

Anyway, as I think I already mentionned, I believe that as long as we
attempt to store dependencies in the repository rather than computiing
them dynamically during the build, we will have problems like this one.

mshinwell commented:
@shindere Just as a note, you need to be careful with "`sed -i`", as on Mac OS X the extension argument is mandatory.  (To not have any backup on that platform you have to pass an empty string as argument after the "`-i`"; however, that form does not seem to work with GNU sed.)

gasche commented:
I will go ahead and merge this (after AppVeyor returns) because I had to cherry-pick this in 4.07 anyway to prepare +rc1, so I think it's good to have the same commit on both sides.

shindere commented:
Mark Shinwell (2018/06/19 07:23 -0700):
> @shindere Just as a note, you need to be careful with "`sed -i`", as
> on Mac OS X the extension argument is mandatory.  (To not have any
> backup on that platform you have to pass an empty string as argument
> after the "`-i`"; however, that form does not seem to work with GNU
> sed.)

Oh, many thanks for having reported this!

@mshinwell: do you believe it would be a problem to have a backup? Am I
correct that it would be a .depend~ file that would be ignored by git
anyway and removed by make clean? Would that be acceptable?

(Assuming we decide to fix this, which I'm not sure is worth it yet --
commments really welcome on this)

xavierleroy commented:
This specific kind of errors could be avoided by using GNU make's `wildcard` function: `$(wildcard *.c)` expands to the empty list if no files matching `*.c` are found.

`set -o pipefail` would be good if only it were portable...

A posteriori `sed -i` would work here but would not work for more complicated cases, such as the `depend:` entry of `byterun/Makefile`.

shindere commented:
Xavier Leroy (2018/06/19 09:53 -0700):
> This specific kind of errors could be avoided by using GNU make's
> `wildcard` function: `$(wildcard *.c)` expands to the empty list if no
> files matching `*.c` are found.

Well would that be enough? I mean, would gcc be happy to be called with
no input file at all?

> A posteriori `sed -i` would work here but would not work for more
> complicated cases, such as the `depend:` entry of `byterun/Makefile`.

Indeed. I guess one portable solution would be to use a fail stamp as in
the testsuite but it becomes really complex and I'm not sure it's worth
it.

One other solution that comes to my mind would be to redirect gcc's
stderr to a file for all its invocations and then check whether that
file is empty. Did I say how ugly I find all this? ;-) :(

xavierleroy commented:
All of your points are well taken.  I agree about the general ugliness.  

Yet another solution: redirect the output of `gcc -MM` to a temp file, then `sed` this temp file.

If we end up doing something ugly like that, let us put it in a shell script in tools/, at least.

shindere commented:
Here is yet another (perhaps less?) ugly approach: for the cases like
byterun, we could have several .depend files and then the sed-based
approach would work. One sed invocation per file. Well I am not in a
hurry about implementing this. :p

shindere commented:
Xavier Leroy (2018/06/19 10:11 -0700):
> All of your points are well taken.  I agree about the general ugliness.  
> 
> Yet another solution: redirect the output of `gcc -MM` to a temp file,
> then `sed` this temp file.

Yeah but as I said we will have to have one temp file per type of object
file, because the different sed invocations replace the same occurrences
of '.o' by different things. Here the real issue is gcc's lack of
flexibility. If one could let it know the name of the output file one
wouldn't need all this stuff.

> If we end up doing something ugly like that, let us put it in a shell
> script in tools/, at least.

Hmm, I am not sure how much stuff we could really factorise in the
script.

shindere commented:
Actually, gcc DOES have the option I was suggesting in my previous
comment! It's `-MT`!

I think we are going to be able to get rid of a few sed
invocations. We'll just have to call gcc on each file rather than on all
of them at the same time but I don't think it's a big deal.
